### PR TITLE
Tests/qr reader overlay

### PR DESCRIPTION
**Pull request by HaGuesto**
If I did changes to any component it was either:
- that I synced it with definitions in the test plan or
- that I got a warning in the tests (mostly memory leak)

### CODE REVIEW

**Comment by HaGuesto** [comment]
[2.0 QrReader/QrReaderOverlay: implement test cases #sprint-jan-19](https://trello.com/c/EKyLiDhR/1069-20-qrreader-qrreaderoverlay-implement-test-cases-sprint-jan-19)

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+  mockImplementationOfQrReader(mockedQrReader, "NoBoxAssociatedWithQrCode");
+  // mock scanning a QR code
+  render(<HeaderMenuContainer />, {
+    routePath: "/bases/:baseId",
+    initialUrl: "/bases/1",
+    mocks: [queryFindNoBoxAssociated],
+    additionalRoute: "/bases/1/boxes/123456",
+  });
+
+  // Open QROverlay
```
Comment (pylipp) [inline react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:57]: Can you put a comment that this is actually test case 3.4.1.1?

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+  await user.click(qrButton);
+
+  // Find Box
+  const findBoxButton = await screen.findByRole("button", { name: /find/i });
+  await user.type(screen.getByRole("textbox"), "123456");
+  await user.click(findBoxButton);
+
+  // error message appears
+  expect(
+    await screen.findByText(/You don't have permission to access this box/i),
```
Comment (pylipp) [inline react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:147]: https://docs.google.com/spreadsheets/d/1sDhSsaVwNxAhGn1VYACDPVepZpvlwTkgy9nmeTaRy2Q/edit?pli=1&disco=AAAApSHIVhs

message should be: "You don't have permission to access Box $identifier"

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+  const qrButton = await screen.findByRole("button", { name: /scan qr code/i });
+  await user.click(qrButton);
+
+  // Find Box
+  const findBoxButton = await screen.findByRole("button", { name: /find/i });
+  await user.type(screen.getByRole("textbox"), "123456");
+  await user.click(findBoxButton);
+
+  // error message appears
+  expect(await screen.findByText(/Box not found for this label/i)).toBeInTheDocument();
```
Comment (pylipp) [inline react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:67]: https://docs.google.com/spreadsheets/d/1sDhSsaVwNxAhGn1VYACDPVepZpvlwTkgy9nmeTaRy2Q/edit?pli=1&disco=AAAApSHIVho

message should be "A box with this label number doesn't exist!"

File: react/src/components/QrReaderOverlay/QrReaderOverlayContainer.tsx [hunk 4]
```diff
@@ -251,18 +207,65 @@ function QrReaderOverlayContainer({
...
+              if (errorCode === "FORBIDDEN") {
+                onScanningDone([
+                  { kind: QrResolverResultKind.NOT_AUTHORIZED, qrCodeValue: qrCode },
+                ]);
+              } else if (errorCode === "BAD_USER_INPUT") {
+                onScanningDone([
+                  { kind: QrResolverResultKind.NOT_BOXTRIBUTE_QR, qrCodeValue: qrCode },
+                ]);
+              } else {
+                onScanningDone([{ kind: QrResolverResultKind.FAIL, qrCodeValue: qrCode }]);
```
Comment (pylipp) [inline react/src/components/QrReaderOverlay/QrReaderOverlayContainer.tsx:236]: You could add another test case with error code `INTERNAL_SERVER_ERROR` to cover this line and the one below.

File: react/src/mocks/components.tsx [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { Result } from "@zxing/library";
+
+export function mockImplementationOfQrReader(mockedQrReader: jest.MockedFunctionDeep<any>, hash: string, isBoxtributeQr:boolean = true) {
```
Comment (pylipp) [inline react/src/mocks/components.tsx:3]: Can you add an explanatory comment here why the mock is required and why it's implemented this way and how to use it in a test.

File: react/src/mocks/functions.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    writable: true,
+    value: jest.fn().mockImplementation((query) => ({
+      matches: returnBool,
+      media: query,
+      onchange: null,
+      addListener: jest.fn(),
+      removeListener: jest.fn(),
+      addEventListener: jest.fn(),
+      removeEventListener: jest.fn(),
+      dispatchEvent: jest.fn(),
```
Comment (pylipp) [inline react/src/mocks/functions.ts:15]: Where do you get the list of these function names from? Same below for the Auth0 mock

**Review by pylipp — CHANGES_REQUESTED** [review]
I compared the test file to the test plan. Some error messages should be different in the implementation, and non-optional test case 3.4.2.4 (Mobile: user scans QR code of same org, but different base with associated box) are not implemented (yet).
Apart from that, glad to see coverage rising :sunglasses:

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx (hunk 3 above)
Comment (HaGuesto) [reply react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:67]: ah I see. I overlooked the comment from Roanna

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx (hunk 2 above)
Comment (HaGuesto) [reply react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:147]: I talked to Roanna about this and we change Box $identifier to this box. I updated it in the Test Plan

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx (hunk 1 above)
Comment (HaGuesto) [reply react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:57]: done

File: react/src/components/QrReaderOverlay/QrReaderOverlayContainer.tsx (hunk 4 above)
Comment (HaGuesto) [reply react/src/components/QrReaderOverlay/QrReaderOverlayContainer.tsx:236]: done

File: react/src/mocks/components.tsx (hunk 5 above)
Comment (HaGuesto) [reply react/src/mocks/components.tsx:3]: done

File: react/src/mocks/functions.ts (hunk 6 above)
Comment (HaGuesto) [reply react/src/mocks/functions.ts:15]: please check the link in the comment: https://jestjs.io/docs/26.x/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx (hunk 2 above)
Comment (HaGuesto) [reply react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx:149]: done

File: react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx (hunk 3 above)
Comment (HaGuesto) [reply react/src/components/QrReaderOverlay/OrReaderOverlay.test.tsx]: done

**Comment by HaGuesto** [comment]
@pylipp Feel free to reveiw it again.

**Review by pylipp — APPROVED** [review]
Nice, thanks for the work :+1:

**Pull request merged**